### ISSUE TITLE
Epoxy Recyclerview displays duplicate items that are not in the dataset

### ISSUE THREAD

**Issue by JayyyR** [airbnb/epoxy#652]
I've run into an issue that seems to be related to some kind of race condition going on in the Epoxy code.

I'm use a PagedListEpoxyController. I do a lot of rapid updates to the data set and Epoxy is usually great for that given the diffing and everything.

But I seem to run into an issue when I have am editing my data set and quickly requesting new model builds. I edit my data set by copying (using kotlin data classes). 

The issue I'm seeing is that the recycler view is displaying items twice even though they aren't in my data set.

The example is for a messaging app
1. I send a message: a new model gets added to my data set with a sending state that is reflected visually on the model view (loader)
2. I immediately send another message, new data gets added again
3. The first message API call returns, the data gets manipulated so the message is no longer in sending state, view gets updated
4. The second message API call returns immediately after, that view gets updated correctly.

The problem is... All the sudden the first set of data that was originally added to my data list is at the bottom of the list. I.E. My first X messages that were loaded when I opened my app are now duplicated below my last 2 messages I just sent. 

I debugged and I'm stumped. I verified that the data set I'm feeding to my controller does not contain the duplicates and thus the recycler view is not reflecting my data set. The behavior is not consistent. It seems to happen when calls to `loadBefore` in my PageKeyedDataSource (Paging library) get triggered. So I can more easily reproduce if I send multiple messages, scroll up to load more data thru paging. I've found the paging library to be kinda finicky so calls to loadBefore can happen a lot.

But even if a lot of calls are being made, I can't figure out why my RecyclerView would be displaying these duplicate items that are NOT in the data set. I verified this with logs...debugging doesn't work because it has to do with some kind of race.

Has this issue come up before? The view displaying things that are not in the data set?

**Comment by TheLester** [comment]
I have similar issue https://github.com/airbnb/epoxy/issues/643 , without using paging library though

**Comment by elihart** [comment]
Thanks for the details, this is quite intriguing and does seem like it could have the same cause as @TheLester 's issue.

Given that you're not feeding duplicate models I'm not sure where the duplicates are coming from. My first guess is that there is an issue in the diffing - possibly a threading issue since it seems to manifest with quick updates.

In `EpoxyControllerAdapter` can you verify that `setModels` is passed the correct list of models? 

Then can you debug the diff results that come back in `onResult`?

**Comment by JayyyR** [comment]
Hey Eli, thanks for the response, I'll test this out and get back to you

**Comment by JayyyR** [comment]
@elihart Through some tough debugging, I've discovered that the models passed into `setModels` in `EpoxyControllerAdapter` is **wrong**.  It contains the duplicates.

The objects are different objects (different references). However their message bodies are the same and their ids ARE the same.

**Comment by JayyyR** [comment]
I can explain how I determined that if that helps

**Comment by JayyyR** [comment]
also verified that when this bug happens, models with duplicate ids get passed to `addModels` in my controller

**Comment by JayyyR** [comment]
@elihart the duplicates are replacing other valid data in the data set. Somehow correct data is getting overwritten with duplicates. I tried filtering out the duplicates, and I can now see that I'm missing correct data

**Comment by elihart** [comment]
Thanks for debugging, that's really helpful to know that the issues comes during model build.

Looking at `PagedListModelCache` I feel like there could be a race condition where the list in the async differ might change while models are being built
```
private fun getOrBuildModel(pos: Int): EpoxyModel<*> {
    modelCache[pos]?.let {
      return it
    }
    return modelBuilder(pos, asyncDiffer.currentList?.get(pos)).also {
      modelCache[pos] = it
    }
  }

  fun getModels(): List<EpoxyModel<*>> {
    (0 until modelCache.size).forEach {
      getOrBuildModel(it)
    }
    lastPosition?.let {
      triggerLoadAround(it)
    }
    @Suppress("UNCHECKED_CAST")
    return modelCache as List<EpoxyModel<*>>
  }
```

it seems like each time `getOrBuildModel` is called there is a risk of getting a different `asyncDiffer.currentList`

Possibly a solution is to call `asyncDiffer.currentList` once at the start of the process and use only that list instance. Possibly some other synchronization needs to happen too.

@yigit I don't expect you to keep maintaining this, so feel free to decline if you're too busy, but I would be interested to hear whether you think this could be an issue.

**Comment by JayyyR** [comment]
Thanks for the reply. Looks like `modelCache` is a privately set variable in `PagedListEpoxyController`. How could i change the behavior of `getOrBuildModel`? 

Also what do you mean once at the start of the process? You mean once per call to `addModels`?

**Comment by elihart** [comment]
This would require changing the Epoxy code with a PR. I imagine it could look like
```
fun getModels(): List<EpoxyModel<*>> {
      val currentList = asyncDiffer.currentList

    (0 until modelCache.size).forEach {pos ->
        modelCache[pos]?.let {
            return@forEach
        }

        modelBuilder(pos, currentList?.get(pos)).also {
            modelCache[pos] = it
        }
    }

    lastPosition?.let {
      triggerLoadAround(it)
    }
    @Suppress("UNCHECKED_CAST")
    return modelCache as List<EpoxyModel<*>>
  }
```

Makes me wonder if we also need to synchronize access to the modelCache since it could be updated from paged list changes as well, but I think that is covered by the requirement that the paged list notify executor uses the same thread as model building
```
* The notify thread of your PagedList (from setNotifyExecutor in the PagedList Builder) must be 
   * the same as this thread. Otherwise Epoxy will crash.
   */
  modelBuildingHandler: Handler = EpoxyController.defaultModelBuildingHandler
```

can you verify that you have your model building and paged list notify executor set up correctly to share the same thread?

**Comment by JayyyR** [comment]
Oh wow, it might not be the same. So model building happens on the UI thread by default? How can I set the notifyExecutor() to be on the same thread? Right now it's set to the same executor as the fetchExecutor

**Comment by JayyyR** [comment]
```
private val networkExecutor = Executors.newFixedThreadPool(5)


val pagedList = PagedList.Builder(pagedAPIDataSource, pagedListConfig)
                .setFetchExecutor(networkExecutor)
                .setNotifyExecutor(networkExecutor)
                .build()
```

^^current code

**Comment by JayyyR** [comment]
hmm I put it on a MainThreadExecutor according to this answer: https://stackoverflow.com/questions/52642246/how-to-get-executor-for-main-thread-on-api-level-28

And still see the issue

**Comment by yigit** [comment]
in theory it does not need to be thread safe because model building is supposed to happen on the same thread  as diffing. That being said, executors only ensure thread pool safety so I'm possibly missing a syncronization there. I'll check.

**Comment by yigit** [comment]
re: https://github.com/airbnb/epoxy/issues/652#issuecomment-452071605
I don't think it is the issue here.
CurrentList on async list differ changes only on the notify thread, which is set by the model cache:
https://github.com/airbnb/epoxy/blob/master/epoxy-paging/src/main/java/com/airbnb/epoxy/paging/PagedListModelCache.kt#L84

And that notify thread is expected to be the same thread as the building thread (i'm assuming that handler is backed by a single thread).

As long as epoxy calls buildModels in the same thread, it should work fine as far as I can see :/.

@JayyyR , can you log the thread build models is called ? Actually if you have a custom epoxy build, it would be nice to get logs for in which thread notify events are called (https://github.com/airbnb/epoxy/blob/master/epoxy-paging/src/main/java/com/airbnb/epoxy/paging/PagedListModelCache.kt#L54) 
You can also log threads in studio's debugger (attach a break point, go to config and change suspend to log)

**Comment by JayyyR** [comment]
@yigit thanks for the info, I'll try to get those logs. So the building thread should be the main thread by default, right?

@elihart  for context, here's what's going on:

I need to edit my models as they chance but I'm using the Paging library. I can't dig into a PagedList and manipulate my data set. So what I'm actually doing is mimicking my data set in a "backing" data set.

So as the PageKeyedDataSource returns with updates to my data (from my API) I also add that data to my backing dataset.

Then if I need to add to the dataset locally or manipulate the data locally, I edit my backing data set. Then I create a whole new PageKeyedDataSource with the (updated) backing data served as the initial data set. (My PageKeyedDataSource will check to see if an initial data set has been passed to the constructor and if so, it will not hit the API for the initial call).

So the problems seem to arise when I manipulate the data set locally (thus editing the backing data set and creating a new PageKeyedDataSource) while also scrolling fast and multiple calls to `loadBefore` are made.

I've verified that my data set is correct. There shouldn't be any issues because if I edit the data locally I pass all the data to a new PageKeyedDataSource as the initial set. And Epoxy should just be getting whatever is in that latest set. It seems like something is screwing up when it gets to the controller.

Hopefully that explains the issue a bit better

**Comment by yigit** [comment]
wait to be clear, you are not modifying an item that you've posted to paged list right?

All of the diffing etc relies on items being immutable so if you are modifying them, diffing wont work.

**Comment by JayyyR** [comment]
right I'm not modifying the item directly, I'm copying the data set objects and changing properties as relevant. Something like 

```
dataset.map { data ->
   data.copy(property = newValue)
}
```

**Comment by yigit** [comment]
there is a delay between submitting a list and seeing it in buildModels (because there is diffing on a background thread by paged list). So it is fine for them to be out of sync as long as it catches up.

Is this a case you are able to repro in a sample ?

**Comment by JayyyR** [comment]
do you mean in a sample project? I can give it a shot. but it might be difficult. The behavior is really inconsistent in my app, but definitely happens enough that its an issue (send a message and scroll up fast and it will happen ~20% of the time)

**Comment by yigit** [comment]
yes that would make it easier. I can try to create a stress test as well but a repro sample as well. as far as I understand, it gets triggered as new items are loaded in the paged list (as user scrolls) while new list is being submitted.

**Comment by JayyyR** [comment]
yes that seems to be the case

**Comment by elihart** [comment]
Thanks for chiming in yigit.

It seems fragile that we assume/hope that the notify thread is the same thread as the building thread .

Is there any reason that we don't just synchronize all access to the model cache? There is a PR to support clearing the cache (https://github.com/airbnb/epoxy/pull/586) which seems like it would need synchronization. Also several people have posted crashes, such as https://github.com/airbnb/epoxy/issues/637

**Comment by JayyyR** [comment]
@elihart 
@yigit 

Happy to report I was able to build a sample out and I can reproduce the error:
https://github.com/JayyyR/PagingEpoxySample

To reproduce, click the Add button - Quickly scroll all the way to the top to trigger loadBefore(). Then quickly scroll back down to the bottom of the list.

What should happen:
You hit add button, a red row gets added, then after ~3 seconds the red color changes to normal.

What happens when it errors:
After ~3 seconds a lot of duplicate, incorrect data gets added below that red row

It might take a few tries to reproduce. If you need to try again, close the activity and reopen to start from scratch so you can re-trigger that first loadBefore call

**Comment by yigit** [comment]
actually something is wrong in the sample (not sure if that is the reason of course).

When I make the following change, it does not reproduce anymore:

```
//create new data source with updated backing data (if it exists)
        pageKeyedDataSource = PageKeyedSampleDataSource(
            initialDataSet = backingDataSet.map { it },
            initialBeforeKey = latestBeforeKey,
            compositeDisposable = dataSourceSubscriptions,
            updateBackingDataCallback = updateBackingData
        )
```
When you add a new item, you are adding the item that is backing the other list, that shifts everything. 
```
backingDataSet.addAll(0, dataToAdd)    
```

Hence your data source returns duplicate data, which is why you see the same items added back

**Comment by yigit** [comment]
@elihart for clearing cache, that can be achieved by posting though of course it will make things async. that can also post and then sync on it.

Relying on the same thread ensures that the backing model is single threaded (only accessed in build models thread which is also where paging notifies [serves as paging's notify cache] ).

Also, `PagedListModelCache` ensures they are the same handler so developer does not need to do anything for that. (unless i'm missing some other callback).

**Comment by JayyyR** [comment]
ohhh, are you saying the `backingDataSet` I'm passing to my PageKeyedSampleDataSource is pointing to the same array list currently? So things get screwed up when I manipulate that list directly afterwards?

So if I do this `.map{it}` it passes a copy and thus editing my `backingDataSet` after doesn't create weird behavior?

**Comment by elihart** [comment]
@yigit I am still hearing from several people that they are getting crashes from a null model in the paged controller (https://github.com/airbnb/epoxy/issues/567)
```
java.lang.NullPointerException: Attempt to invoke virtual method 'void com.airbnb.epoxy.EpoxyModel.addTo(com.airbnb.epoxy.EpoxyController)' on a null object reference
at com.airbnb.epoxy.EpoxyController.add(EpoxyController.java:465)
at com.airbnb.epoxy.paging.PagedListEpoxyController.addModels(PagedListEpoxyController.kt:78)
at com.airbnb.epoxy.paging.PagedListEpoxyController.buildModels(PagedListEpoxyController.kt:70)
at com.airbnb.epoxy.EpoxyController$1.run(EpoxyController.java:267)
```

which makes me wonder if this is related, and if synchronizing may be necessary.

You previously said `It is important that the notify thread of PagedList is the same as the buildModels thread of EpoxyController.` - and this is the threading setup that I felt like was fragile and not enforceable.

People are doing workarounds like this: https://github.com/chrisbanes/tivi/commit/a93afc29e313b311d3e7c6bbf1855f8f0e91c14a

**Comment by elihart** [comment]
@JayyyR did that change to how you handle your `backingDataSet` fix the issue?

**Comment by JayyyR** [comment]
@elihart it does seem to have fixed my issue. I'm still not 100% clear on what happens without the fix but do you think my understanding above makes any sense?

It seems this is related to how the paging library works

**Comment by yigit** [comment]
@elihart i'll check the CL in detail but i don't think sync will help.
At a given time T, there are two lists,
- the last list submitted by the async differ (source)
- the model, based on that list (dst)

Every time `source` sends a update, we have to update the `dst` accordingly (they are just changes or additions).
When a new list (a new `source`) comes, we receive a diff between the old `source` and the new `source` and then invalidate the `dst` accordingly.

Now, async page differ only guarantees to makes these syncronized between the notify and compute threads. 
So either we need to keep the model on that thread OR use message passing to move notifications from the async's notify thread to model's build thread. Without that kind of mechanism, it is not going to be trivial to handle all various updates that might arrive.

I'll check the sample in #567 to see if i can figure out what is going on. But the issue in this ticket is what i've mentioned here: https://github.com/airbnb/epoxy/issues/652#issuecomment-452554171
It is a bad data source that is not returning consistent data, which confuses the diffing.

**Comment by yigit** [comment]
i think i know what is going on there:
https://github.com/airbnb/epoxy/issues/567#issuecomment-453776573

Though I'm not sure if this is intended behavior or a bug.

**Comment by elihart** [comment]
Some synchronization issues have been fixed in the 3.2.0 release, let me know if there are any further problems.
